tests/testthat/test-type-misc.R

```diff
@@ -29,6 +29,133 @@ test_that("`numeric_version` falls back to base methods", {
   expect_identical(vec_c(x, y), z)
 })
 
+test_that("`numeric_version` has an equality, comparison, and order proxy", {
+  numeric_row <- function(...) {
+    out <- list2(...)
+    out <- map(out, as.integer)
+    names(out) <- paste0("...", seq_len(8L))
+    new_data_frame(out, n = 1L)
+  }
+
+  x <- numeric_version(c("1.2-3", "1.21.1", "3", "2.21.0.9000", "0.5.01"))
+
+  expect <- vec_rbind(
+    numeric_row(1, 2, 3, 0, 0, 0, 0, 0),
+    numeric_row(1, 21, 1, 0, 0, 0, 0, 0),
+    numeric_row(3, 0, 0, 0, 0, 0, 0, 0),
+    numeric_row(2, 21, 0, 9000, 0, 0, 0, 0),
+    numeric_row(0, 5, 1, 0, 0, 0, 0, 0)
+  )
+
+  expect_identical(vec_proxy_equal(x), expect)
+  expect_identical(vec_proxy_compare(x), expect)
+  expect_identical(vec_proxy_order(x), expect)
+})
+
+test_that("`numeric_version` proxy works with empty vectors", {
+  x <- numeric_version(character())
+
+  expect <- vec_rep(list(integer()), times = 8L)
+  names(expect) <- paste0("...", seq_len(8L))
+  expect <- new_data_frame(expect, n = 0L)
+
+  expect_identical(vec_proxy_equal(x), expect)
+})
+
+test_that("`numeric_version` proxy handles pseudo-`NA`", {
+  numeric_row <- function(...) {
+    out <- list2(...)
+    out <- map(out, as.integer)
+    names(out) <- paste0("...", seq_len(8L))
+    new_data_frame(out, n = 1L)
+  }
+
+  x <- numeric_version(c("1_1", "1.2", NA), strict = FALSE)
+
+  expect <- vec_rbind(
+    numeric_row(NA, NA, NA, NA, NA, NA, NA, NA),
+    numeric_row(1, 2, 0, 0, 0, 0, 0, 0),
+    numeric_row(NA, NA, NA, NA, NA, NA, NA, NA)
+  )
+
+  expect_identical(vec_proxy_equal(x), expect)
+  expect_identical(vec_proxy_compare(x), expect)
+  expect_identical(vec_proxy_order(x), expect)
+})
+
+test_that("`numeric_version` works with functions using the equality proxy", {
```

Comment:
Also add tests comparing versions with 8 components to versions with fewer components?